### PR TITLE
712 username as identifier for avatars and in chat comment function

### PR DESCRIPTION
**Pull request by SarahAugustinowski**
(This PR is the new version for #743 which was closed because of issues with SEENewWorld)

With this you can now set a username (playername) as a avatar tag above your player and for the chat function.

I am not sure if the NullReferenceException for the FaceCam was there before my implementation or its effected by mine.

### CODE REVIEW

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerName.cs:59]: Why does this need to be called every frame? Shouldn't it suffice to set the text once?

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerName.cs:81]: The method name seems excessively long—maybe something like `SendPlayerNameToClients` can be used instead?

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerName.cs:79]: Missing documentation.

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerName.cs:66]: The description seems off—doesn't this just render the player name instead of sending it to clients? Maybe I'm misunderstanding how the `ClientRpc` and `ServerRpc` attributes work, though.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:10]: Missing class-level documentation.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:15]: Instead of defining this delegate, consider using `Action` instead.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:61]: Extraneous newlines.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:82]:
```suggestion
                playerName.Description = "Username which will be used as avatar-tag and elsewhere.";
```

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:128]: The formatting here is weird (e.g., there shouldn't be a newline before calling a function).

Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:128]:
```suggestion
                    ShowNotification.Error
                       ("Username Error",
                        "Username needs to be at least one character long.");
```

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:139]: Instead of using the `errorOccurred` variable, consider unnesting the block containing the username check and just putting the contents of this if-block in the `else` block above.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:149]: Should use primitive instead of boxed type:
```suggestion
            return !string.IsNullOrWhiteSpace(playername);
```

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:151]: Extraneous newline

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:162]: Missing EOF newline.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:59]: I think these can be set to `private`.

**Review by falko17 — CHANGES_REQUESTED** [review]
Thanks for the PR! I've added some questions and comments below.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (SarahAugustinowski) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:15]: I used this way because it is the same as in NetworkPropertyDialog. Should i change both classes?

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (SarahAugustinowski) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:128]: copied from NetworkPropertyDialog and adapt it, but will change same both ;-)

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (SarahAugustinowski) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:139]: I used this way because it is the same as in NetworkPropertyDialog. Should i change both classes?

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Tools/Playername/PlayerName.cs:85]: i deleted the serverRpcParam because it was default and is not necessary to add then

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:18]: Yes, please do.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:138]: Yes, please change both (at least in my opinion). I don't really see the point of the variable and I don't know why the code was this way in `NetworkPropertyDialog`.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:62]: I'm not sure if these are maybe used from outside anywhere in `NetworkPropertyDialog`. Please check that first, and if it isn't the case, then you can set those ones to private too. I think the ones in `UserPropertyDialog` aren't used from outside at least, so these can be set to private in any case.

File: Assets/SEE/Game/Worlds/PlayerSpawner.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:74]: I am not familiar with yield and coroutines. But i now think this is totally wrong. :) Shouldn't it be like this (similiar to the networkManager check):
```
while (ReferenceEquals(networkConfig, null))
{
    Debug.LogError("Network configuration not found");
    networkConfig = FindObjectOfType<SEE.Net.Network>();
    yield return null;
}
```
My understanding would be that routine pause and try to find the missing networkConfig later again?

Comment (falko17) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:74]: Yes, something like this does look a lot more reasonable. I'd even recommend using `async` functionality here instead:
```csharp
await UniTask.WaitUntil(() => !ReferenceEquals(networkConfig = FindObjectOfType<SEE.Net.Network>(), null));
```
This should have the same effect as the coroutine approach above (checking every frame whether the Network object is present). I don't know if this is really necessary at all, though—if we do keep it, we should remove the error message (like in the `async` example above), otherwise the message will appear every frame.

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Tools/Playername/PlayerName.cs:62]: The rendering it self is done in `RenderNetworkPlayername` . Clients call this method so the Server recieves the name and send it to all other clients. Thats how I understood this https://docs-multiplayer.unity3d.com/netcode/current/advanced-topics/message-system/serverrpc/, but might also be not 100% right.

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Tools/Playername/PlayerName.cs:55]: If i put it into `Start` it doesn't work - not sure why.

Comment (falko17) [reply Assets/SEE/Tools/Playername/PlayerName.cs:55]: Please try to investigate why this is the case. Anything that we can move outside `Update()`, we should. By "it doesn't work", do you mean nothing happens, or some sort of error is thrown/appears in the console?

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (falko17) [reply Assets/SEE/Tools/Playername/PlayerName.cs:62]: Okay, thanks for the explanation. I think I'll defer to @koschke here, since I don't really have any experience with the networking stuff in SEE.

**Comment by koschke** [comment]
@SarahAugustinowski What is the current status of this pull request? There hasn't been any recent activity.

Please note that we will soon have new avatars. @Yvo02 is currently working on this on another branch.

**Comment by SarahAugustinowski** [comment]
> @SarahAugustinowski What is the current status of this pull request? There hasn't been any recent activity.
> 
> Please note that we will soon have new avatars. @Yvo02 is currently working on this on another branch.

Sorry for that, I will focus on the PR this weekend again.

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Tools/Playername/PlayerName.cs]: i fixed this.

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (SarahAugustinowski) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs]: fixed this

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (SarahAugustinowski) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs]: for the UserPropertyDialog i changed it. 

For NetworkPropertyDialog i let this unchanged because there a mutiple checks that can cause an error and i dont know how i can do it better without a variable that changes when some where a error occured.

**Comment by SarahAugustinowski** [comment]
I made some recent changes. You might take a second look @falko17 @koschke

File: Assets/SEE/Game/Worlds/PlayerSpawner.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:74]: This is something i am still struggeling with. To use your code snippet i need to make the whole Coroutine a UniTask and remove everything which is like "yield return null" as i understand. But i didn't get it running without freezing :/

**Comment by Yvo02** [comment]
> I made some recent changes. You might take a second look @falko17 @koschke

Should I wait with the PR for #733 until this has been merged into the master? There are changes to the avatar prefabs (male1-3/female1-3), which must also be applied to the new avatars.

**Comment by SarahAugustinowski** [comment]
> > I made some recent changes. You might take a second look @falko17 @koschke
> 
> Should I wait with the PR for #733 until this has been merged into the master? There are changes to the avatar prefabs (male1-3/female1-3), which must also be applied to the new avatars.

I only added a TextMeshPro above the player and put PlayerName and FaceCamera as script on it. 
grafik

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:13]:
```suggestion
    /// Manager for Server to store playernames of the connected clients
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:17]:
```suggestion
        /// <summary>
        /// Dictionary to store player names with their client IDs as keys.
        /// </summary>
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:27]: Unnecessary empty line:
```suggestion
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:23]:
```suggestion
        /// Sets the name of the given <paramref name="clientId"/> to <paramref name="playerName"/>.
        /// </summary>
        /// <param name="clientId">client id of the connected client</param>
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:40]: Unnecessary empty line:
```suggestion
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:44]:
```suggestion
        /// Returns the playerName for the given <paramref name="clientId"/>.
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:54]:
```suggestion
        /// Removes dictionary entry for the given <paramref name="clientId"/>.
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:62]: No need for the if condition here:
```suggestion
            playerNames.Remove(clientId);
```

File: Assets/SEE/Tools/Playername/PlayerNameManager.cs
Comment (falko17) [inline Assets/SEE/Tools/Playername/PlayerNameManager.cs:66]: Unnecessary empty lines:
```suggestion
    }
```

File: Assets/SEE/UI/OpeningDialog.cs
Comment (falko17) [inline Assets/SEE/UI/OpeningDialog.cs]: Should this code have been removed?

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs:87]:
```suggestion
            // to be closed until the input is valid. That is why we will handle the
```

File: Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs
Comment (falko17) [reply Assets/SEE/UI/PropertyDialog/UserPropertyDialog.cs]: Looks good to me!

File: Assets/SEE/Game/Worlds/PlayerSpawner.cs
Comment (falko17) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:74]: Hmm, okay. Just to check, does the following work? (Untested, just to give you an idea of what it would look like to make this method fully async)

## Method itself
```csharp
networkConfig = FindObjectOfType<SEE.Net.Network>();
await UniTask.WaitUntil(() => !ReferenceEquals(networkConfig = FindObjectOfType<SEE.Net.Network>(), null));
await UniTask.WaitUntil(() => !ReferenceEquals(networkManager = NetworkManager.Singleton, null));

// Terminate this function if not run by the server, but first set client playername for chat-function.
if (!networkManager.IsServer)
{
    dissonanceComms = FindObjectOfType<DissonanceComms>();
    dissonanceComms.LocalPlayerName = networkConfig.PlayerName;
    return;
}

// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
// The following code will be executed only on the server.
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
// Wait until Dissonance is created
await UniTask.WaitUntil(() => !ReferenceEquals(dissonanceComms = FindObjectOfType<DissonanceComms>(), null));
dissonanceComms.LocalPlayerName = networkConfig.PlayerName;

// The callback to invoke once a client connects. This callback is only
// ran on the server and on the local client that connects. We want
// to spawn a player whenever a client connects.
networkManager.OnClientConnectedCallback += Spawn;
networkManager.OnClientDisconnectCallback += ClientDisconnects;
// Spawn the local player. This code is executed by the server.
Spawn(networkManager.LocalClientId);
```

## Method signature
```csharp
private async UniTaskVoid SpawnPlayer()
```

## Method usage
AFAICT, this method is only used in `OnEnable`, but any other usages would need to be replaced as well:
```csharp
// Previously: StartCoroutine(SpawnPlayerCoroutine());
SpawnPlayerCoroutine().Forget();
```

---

If this still doesn't work (e.g., freezes), we could open a follow-up issue for this instead of doing it in this PR if that takes too much time. There are some other places in SEE that need to be updated to be async anyway.

**Review by falko17 — CHANGES_REQUESTED** [review]
Thanks for the changes, there's only some minor formatting/documentation suggestions left now.

**Comment by koschke** [comment]
Looks like there is still some work to do. Please do not forget to merge the master branch into your branch.

File: Assets/SEE/UI/OpeningDialog.cs
Comment (SarahAugustinowski) [reply Assets/SEE/UI/OpeningDialog.cs]: yes, it is outdated

File: Assets/SEE/Game/Worlds/PlayerSpawner.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:74]: Looks similar to one of my attempts. Tried this again but not working. :( It freezes and can't set up playername, it says dissonance is started already and then the playername can't be set.

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Tools/Playername/PlayerName.cs:73]: @koschke Can this be resolved?

File: Assets/SEE/Game/Worlds/PlayerSpawner.cs
Comment (falko17) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:74]: Oh well, not sure what causes this—I created a follow-up issue for this now (#762) as it's not essential for this PR, so we can close this thread.

**Review by koschke — APPROVED** [review]
All is fine now. I fixed the remaining issues myself.

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (koschke) [reply Assets/SEE/Tools/Playername/PlayerName.cs]: I fixed it.

**Review by koschke — APPROVED** [review]
All resolved.

File: Assets/SEE/Game/Worlds/PlayerSpawner.cs
Comment (falko17) [inline Assets/SEE/Game/Worlds/PlayerSpawner.cs:102]: Missing documentation.

Comment (falko17) [inline Assets/SEE/Game/Worlds/PlayerSpawner.cs:102]: Currently, we are surrounding every call of this method with an `#if DEBUG`. Instead, [Unity documentation recommends](https://docs.unity3d.com/Manual/UnderstandingPerformanceGeneralOptimizations.html) using the `Conditional` attribute on this `Log` method. That way, we don't have to wrap every call with these preprocessor directives and only have to do it once.

File: Assets/SEE/UI/PropertyDialog/NetworkPropertyDialog.cs
Comment (falko17) [inline Assets/SEE/UI/PropertyDialog/NetworkPropertyDialog.cs:74]: AE vs BE:
```suggestion
        /// The password to protect the server from unauthorized clients
```

**Review by falko17 — APPROVED** [review]
Just three minor comments left, so I'll approve already.

File: Assets/SEE/Game/Worlds/PlayerSpawner.cs
Comment (koschke) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:107]: Added.

Comment (koschke) [reply Assets/SEE/Game/Worlds/PlayerSpawner.cs:107]: Cool feature. Thanks for the hint. It's used now.

File: Assets/SEE/Tools/Playername/PlayerName.cs
Comment (SarahAugustinowski) [reply Assets/SEE/Tools/Playername/PlayerName.cs]: Thanks! Also for every other fix :)

**Pull request merged**
